python/cudf/cudf/core/frame.py

```diff
@@ -3337,18 +3329,240 @@ def _reindex(
 
         return self._mimic_inplace(result, inplace=inplace)
 
+    def _unaryop(self, op):
+        data_columns = (col.unary_operator(op) for col in self._columns)
+        data = zip(self._column_names, data_columns)
+        return self.__class__._from_table(Frame(data, self._index))
+
+    def _binaryop(
+        self,
+        other: T,
+        fn: str,
+        fill_value: Any = None,
+        reflect: bool = False,
+        *args,
+        **kwargs,
+    ) -> Frame:
+        raise NotImplementedError
+
+    @classmethod
+    def _colwise_binop(
+        cls,
+        operands: Dict[Optional[str], Tuple[ColumnBase, Any, bool, Any]],
+        fn: str,
+    ):
+        """Implement binary ops between two frame-like objects.
+
+        Binary operations for Frames can be reduced to a sequence of binary
+        operations between column-like objects. Different types of frames need
+        to preprocess different inputs, so subclasses should implement binary
+        operations as a preprocessing step that calls this method.
+
+        Parameters
+        ----------
+        operands : Dict[Optional[str], Tuple[ColumnBase, Any, bool, Any]]
+            A mapping from column names to a tuple containing left and right
+            operands as well as a boolean indicating whether or not to reflect
+            an operation and fill value for nulls.
+        fn : str
+            The operation to perform.
+
+        Returns
+        -------
+        Frame
+            A subclass of Frame constructed from the result of performing the
+            requested operation on the operands.
+        """
+
+        # Now actually perform the binop on the columns in left and right.
+        output = {}
+        for (
+            col,
+            (left_column, right_column, reflect, fill_value),
+        ) in operands.items():
+            if right_column is cudf.NA:
+                right_column = cudf.Scalar(
+                    right_column, dtype=left_column.dtype
+                )
+            elif not isinstance(right_column, ColumnBase):
+                right_column = left_column.normalize_binop_value(right_column)
+
+            fn_apply = fn
+            if fn == "truediv":
+                # Decimals in libcudf don't support truediv, see
+                # https://github.com/rapidsai/cudf/pull/7435 for explanation.
+                if is_decimal_dtype(left_column.dtype):
+                    fn_apply = "div"
+
+                # Division with integer types results in a suitable float.
+                truediv_type = {
```

Comment:
nit: could move this to global scope or utils

Reply:
In the past I think we've been overzealous about trying to define function-local constants/dicts like this for performance for things that are relatively negligible. Binops take on the order of milliseconds, whereas defining this dict takes <1 microsecond. IMO the improved readability of having this defined right here (especially since it's not used anywhere else) is more valuable than saving that tiny amount of time. We can always revisit this decision in the future if we actually notice this in profiles. Thoughts?

Reply:
👍